text/2021-10-08-monitoring-metrics-about-chaos-mesh.md

```diff
@@ -0,0 +1,241 @@
+These metrics are not easy to reflect the overall appearance of chaos-mesh system,
+so we need more metrics to improve the observability.
+
+As for the expected outcome, below is several metrics about logic pattern and
+performance should be implemented and exposed on `/metrics` HTTP endpoint:
+
+chaos-controller-manager:
+
+- time histogram for each `Reconcile()` in `Reconciler`
+- count of Chaos Experiments, Schedule, and Workflow
+- count of the emitted event by chaos-controller-manager
+- common metrics of grpc-client
+- metrics for kubernetes webhook
+
+chaos-daemon:
+
+- common metrics of grpc-server (already provided)
+- count of processes controlled by `bpm` (background process manager)
+- count of iptables/ipset/tc rules
+
+chaos-dashboard:
+
+- time histogram for each HTTP query
+- count for archived object
+- time histogram for archive reconciler
+
+## Detailed design
+
+### Metrics Plan
+
+The metrics design is as follows. It should be noted that the implementations
+of `chaos_controller_manager_chaos_experiments_count` and
+`chaos_daemon_grpc_server_handling_seconds` have been provided, and they are
+called `chaos_mesh_experiments` and `grpc_server_handling_seconds`. It is just
+hoping to modify the names to standardize naming. More about naming, see this
+guideline: [Metric and label naming](https://prometheus.io/docs/practices/naming/).
+
+<!-- markdownlint-disable MD013 -->
+| Name                                                  | Description                                            | Type         | Label                       | Buckets                      |
+|-------------------------------------------------------|--------------------------------------------------------|--------------|-----------------------------|------------------------------|
+| chaos_controller_manager_reconcile_duration_seconds   | time histogram for each `Reconcile()` in `Reconciler`  | HistogramVec | type                        | DefBuckets                   |
+| chaos_controller_manager_chaos_schedule_count         | currrent count of Schedule                             | GaugeVec     | namespace                   | /                            |
+| chaos_controller_manager_chaos_workflow_count         | currrent count of Workflow                             | GuageVec     | namespace                   | /                            |
+| chaos_controller_manager_emitted_event_count          | count of the emitted event by chaos-controller-manager | GaugeVec     | type, reason                | /                            |
+| chaos_controller_manager_grpc_client_handling_seconds | common metrics of grpc-client                          | HistogramVec | code, method, service, type | DefBuckets                   |
+| chaos_controller_manager_webhook_duration_seconds     | metrics for kubernetes webhook - duration              | HistogramVec | type, allowed               | WebhookDurationBuckets       |
```

Comment:
Let's be more specific about the metric name. Duration for what? If it is for http requests then `chaos_controller_manager_webhook_http_request_duration_seconds` is better